tests/test_smc.py

```diff
@@ -0,0 +1,164 @@
+        combined_x = np.concatenate([solvent_x[:-num_ligand_atoms], ligand_x], axis=0)
+        combined_box = solvent_samples[choice_idx].box
+        all_xbs.append((combined_x, combined_box))
+    return all_xbs
+
+
+def test_smc():
+    """
+    Generate samples from the equilibrium distribution at lambda=1
+    """
+
+    seed = 2021
+    np.random.seed(seed)
+
+    mol, torsion_idxs = test_ligands.get_biphenyl()
+
+    @jax.jit
+    def get_torsion(x_l):
+        ci = x_l[torsion_idxs[:, 0]]
+        cj = x_l[torsion_idxs[:, 1]]
+        ck = x_l[torsion_idxs[:, 2]]
+        cl = x_l[torsion_idxs[:, 3]]
+        # last [0] is used to return from a length-1 array
+        return bonded.signed_torsion_angle(ci, cj, ck, cl)[0]
+
+    ff = get_ff_am1cc()
+
+    ligand_masses = np.array([a.GetMass() for a in mol.GetAtoms()])
+    num_ligand_atoms = len(ligand_masses)
+
+    temperature = 300.0
+    pressure = 1.0
+
+    ubps, params, masses, coords, box = enhanced.get_solvent_phase_system(mol, ff)
+
+    cache_path = "test_smc_cache.pkl"
+    if not os.path.exists(cache_path):
+        print("Generating cache")
+
+        n_solvent_samples = 100  # should be maybe 1000
+        print(f"generating {n_solvent_samples} solvent samples")
+        solvent_xvbs = generate_solvent_samples(
+            coords, box, masses, ubps, params, temperature, pressure, seed, n_solvent_samples
+        )
+
+        n_ligand_batches = 1200  # should be 30k
+        print(f"generating ligand samples")
+        ligand_samples, ligand_log_weights = generate_ligand_samples(n_ligand_batches, mol, ff, temperature, seed)
+
+        with open(cache_path, "wb") as fh:
+            pickle.dump([solvent_xvbs, ligand_samples, ligand_log_weights], fh)
+
+    with open(cache_path, "rb") as fh:
+        print("Loading cache")
+        solvent_xvbs, ligand_samples, ligand_log_weights = pickle.load(fh)
+
+    n_endstate_samples = 5000
+    all_xbs = generate_endstate_samples(
+        n_endstate_samples, solvent_xvbs, ligand_samples, ligand_log_weights, num_ligand_atoms
+    )
+
+    # plot torsions at the end-states
+    end_state_torsions = []
+    for xb in all_xbs:
+        x = xb[0]
+        x_l = x[:-num_ligand_atoms]
+        end_state_torsions.append(get_torsion(x_l))
+
+    # plot histogram using asciiplotlib
+    fig = apl.figure()
+    fig.hist(*np.histogram(end_state_torsions, bins=25, range=(-np.pi, np.pi)))
+    fig.show()
+
+    bound_impls = []
+    for u, p in zip(ubps, params):
+        bound_impls.append(u.bind(p).bound_impl(np.float32))
+
+    # propagate with NPTMove
+    n_steps = 1000
+    npt_lamb = 1.0
+    seed = int(time.time())
+
+    mover = NPTMove(ubps, npt_lamb, masses, temperature, pressure, n_steps, seed)
+
+    U_fn = functional.construct_differentiable_interface_fast(ubps, params)
+
+    for xi, bi in all_xbs:
+        # (ytz): note to jfass, change v0 to not zeros_like
+        xvb = CoordsVelBox(xi, np.zeros_like(xi), bi)
```

Comment:
Something like 

```python
from md.thermostat.utils import sample_velocities
v0 = sample_velocities(masses * unit.amu, temperature)
```

?

Reply:
I think there's some nuance here, with BAOAB I think we'd still need to adjust the velocities by half a step before we can do `set_v_t(sample_velocities())` - I was hoping we can get this detail correct.

Reply:
Good call.

Illustration of this issue in a 1D system, at a large stepsize: https://gist.github.com/maxentile/1bd50720fe09f0541be0598f2fe37ec9 

A possible workaround appears to be something like `half_step_back_v = equilibrium_v - 0.5 * dt * reduced_force(equilibrium_x) / masses`

**1. _Equilibrium_ distribution**
![image](https://user-images.githubusercontent.com/5759036/144247010-a34033e4-4d83-4f35-9f07-2d29b27b0f43.png)
(Note that x and v are independent!)

**2. _Steady-state_ distribution sampled by our Langevin integrator**
![image](https://user-images.githubusercontent.com/5759036/144246128-cfcdaa58-dc58-4cc7-b151-019d50fd86bd.png)
(Note that x and v marginals are still accurate, but x and v are no longer independent!)

**3. Distribution sampled by running our Langevin integrator for 1 step, _initialized with (x, v) from equilibrium_**
![image](https://user-images.githubusercontent.com/5759036/144246476-561156ce-b15b-4555-ae5d-776eee7189f0.png)
(Note that x marginal is now pretty inaccurate!)

**4. Distribution sampled by running our Langevin integrator for 1 step, _initialized with x from equilibrium, v adjusted by half a step_**
![image](https://user-images.githubusercontent.com/5759036/144246649-7ba1c320-b469-4567-a7ce-39bf01c4cecf.png)
(Note that x marginal is mostly recovered!)

------------

Other workarounds include:
* Running for more MD steps
* Running at a moderate stepsize